components/service/glean/src/main/java/mozilla/components/service/glean/DatetimeMetricType.kt

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean
+
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.launch
+import mozilla.components.service.glean.storages.DatetimesStorageEngine
+import mozilla.components.service.glean.utils.parseISOTimeString
+import mozilla.components.service.glean.utils.truncateDate
+import mozilla.components.support.base.log.logger.Logger
+import java.util.Calendar
+import java.util.Date
+
+/**
+ * This implements the developer facing API for recording datetime metrics.
+ *
+ * Instances of this class type are automatically generated by the parsers at build time,
+ * allowing developers to record values that were previously registered in the metrics.yaml file.
+ */
+data class DatetimeMetricType(
+    override val disabled: Boolean,
+    override val category: String,
+    override val lifetime: Lifetime,
+    override val name: String,
+    override val sendInPings: List<String>,
+    val timeUnit: TimeUnit = TimeUnit.Nanosecond
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = listOf("metrics")
+
+    private val logger = Logger("glean/DatetimeMetricType")
+
+    /**
+     * Set a datetime value.
+     *
+     * @param value The [Date] value to set. If not provided, will record the current time.
+     */
+    fun set(value: Date = Date()) {
+        // TODO report errors through other special metrics handled by the SDK. See bug 1499761.
+        if (!shouldRecord(logger)) {
+            return
+        }
+
+        val truncatedValue = truncateDate(value, timeUnit)
+
+        Dispatchers.API.launch {
+            // Delegate storing the datetime to the storage engine.
+            DatetimesStorageEngine.set(
+                this@DatetimeMetricType,
+                truncatedValue
+            )
+        }
+    }
+
+    /**
+     * Set a datetime value.
+     *
+     * This is provided as an internal-only function so that we can test that timezones
+     * are passed through correctly.  The normal public interface uses [Date] objects which
+     * are always in the local timezone.
+     *
+     * @param value The [Calendar] value to set. If not provided, will record the current time.
+     */
+    internal fun set(value: Calendar) {
```

Comment:
Is this something that we can mark with `@VisibleForTesting(otherwise = VisibleForTesting.NONE)`?

Reply:
Good idea.